File: Client_Root/Client/Assets/Scripts/Network/JSONHelper.cs
using UnityEngine;
using System.Collections.Generic;

public static class JSONHelper
{
    #region AddField
    public static void AddField(JSONObject jsonObject, string strFieldName, int value)
    {
        jsonObject.AddField(strFieldName, value);
    }

    public static void AddField(JSONObject jsonObject, string strFieldName, long value)
    {
        jsonObject.AddField(strFieldName, value);
    }

    public static void AddField(JSONObject jsonObject, string strFieldName, float value)
    {
        jsonObject.AddField(strFieldName, value);
    }

    public static void AddField(JSONObject jsonObject, string strFieldName, string value)
    {
        jsonObject.AddField(strFieldName, value);
    }

    public static void AddField(JSONObject jsonObject, string strFieldName, JSONObject value)
    {
        jsonObject.AddField(strFieldName, value);
    }

    public static void AddField<T>(JSONObject jsonObject, string strFieldName, List<T> value) where T : IJSONObjectable
    {
        JSONObject arr = new JSONObject(JSONObject.Type.ARRAY);

        foreach(T data in value)
        {
            arr.Add(data.GetJSONObject());
        }

        arr.AddField(strFieldName, arr);

        jsonObject.AddField(strFieldName, arr);
    }

    public static void AddField(JSONObject jsonObject, string strFieldName, Dictionary<int, string> value)
    {
        JSONObject obj = new JSONObject(JSONObject.Type.OBJECT);

        obj.keys = new List<string>();
        obj.list = new List<JSONObject>();

        foreach(KeyValuePair<int, string> kv in value)
        {
            obj.keys.Add(kv.Key.ToString());
            obj.list.Add(JSONObject.CreateStringObject(kv.Value));
        }

        jsonObject.AddField(strFieldName, obj);
    }
    #endregion

    #region GetField
    public static bool GetField(JSONObject jsonObject, string strFieldName, ref int value)
    {
        if(!jsonObject.HasField(strFieldName))
        {
            Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
            return false;
        }

        if(!jsonObject.GetField(strFieldName).IsNumber)
        {
            Debug.LogWarning("Data type is invalid! It's not number");
            return false;
        } 

        return jsonObject.GetField(ref value, strFieldName);
    }

    public static bool GetField(JSONObject jsonObject, string strFieldName, ref long value)
    {
        if(!jsonObject.HasField(strFieldName))
        {
            Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
            return false;
        }

        if(!jsonObject.GetField(strFieldName).IsNumber)
        {
            Debug.LogWarning("Data type is invalid! It's not number");
            return false;
        } 

        return jsonObject.GetField(ref value, strFieldName);
    }

    public static bool GetField(JSONObject jsonObject, string strFieldName, ref ulong value)
    {
        if(!jsonObject.HasField(strFieldName))
        {
            Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
            return false;
        }

        if(!jsonObject.GetField(strFieldName).IsString)
        {
            Debug.LogWarning("Data type is invalid! It's not string");
            return false;
        }

        string strValue = "";
        if(!jsonObject.GetField(ref strValue, strFieldName))
        {
            Debug.LogWarning("Failed to GetField");
            return false;
        }
        value = uint.Parse(strValue);

        return true;
    }

    public static bool GetField(JSONObject jsonObject, string strFieldName, ref float value)
    {
        if(!jsonObject.HasField(strFieldName))
        {
            Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
            return false;
        }

        if(!jsonObject.GetField(strFieldName).IsNumber)
        {
            Debug.LogWarning("Data type is invalid! It's not number");
            return false;
        } 

        return jsonObject.GetField(ref value, strFieldName);
    }

    public static bool GetField(JSONObject jsonObject, string strFieldName, ref string value)
    {
        if(!jsonObject.HasField(strFieldName))
        {
            Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
            return false;
        }

        if(!jsonObject.GetField(strFieldName).IsString)
        {
            Debug.LogWarning("Data type is invalid! It's not string");
            return false;
        } 

        return jsonObject.GetField(ref value, strFieldName);
    }

    public static bool GetField(JSONObject jsonObject, string strFieldName, JSONObject value)
    {
        if(!jsonObject.HasField(strFieldName))
        {
            Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
            return false;
        }

        if(!jsonObject.GetField(strFieldName).IsObject)
        {
            Debug.LogWarning("Data type is invalid! It's not object");
            return false;
        } 

        value = jsonObject.GetField(strFieldName);

        return value != null ? true : false;
    }

    public static bool GetField<T>(JSONObject jsonObject, string strFieldName, List<T> listValue) where T : IJSONObjectable, new()
    {
        if(!jsonObject.HasField(strFieldName))
        {
            Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
            return false;
        }

        if(!jsonObject.GetField(strFieldName).IsArray)
        {
            Debug.LogWarning("Data type is invalid! It's not array");
            return false;
        }
            
        foreach(JSONObject element in jsonObject.list)
        {
            T value = new T();
            value.SetByJSONObject(element);

            listValue.Add(value);
        }

        return true;
    }

    public static bool GetField(JSONObject jsonObject, string strFieldName, Dictionary<int, string> dicValue)
    {
        if(!jsonObject.HasField(strFieldName))
        {
            Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
            return false;
        }

        if(!jsonObject.GetField(strFieldName).IsObject)
        {
            Debug.LogWarning("Data type is invalid! It's not object");
            return false;
        }

        JSONObject field = jsonObject.GetField(strFieldName);
        for (int i = 0; i < field.list.Count; i++)
        {
            dicValue.Add(int.Parse(field.keys[i]), field.list[i].str);      
        }

        return true;
    }
    #endregion
}

File: Server_Root/Server/CommonSources/Message/JSONHelper.cpp
#include "stdafx.h"
#include "JSONHelper.h"
#include "IJSONObjectConvertible.h"

JSONHelper::JSONHelper()
{
}


JSONHelper::~JSONHelper()
{
}

#pragma region AddField
void JSONHelper::AddField(Value* pJsonObj, Document* pDocument, const char* pCharFieldName, __int32 value)
{
	pJsonObj->AddMember<__int32>(pCharFieldName, value, pDocument->GetAllocator());
}

void JSONHelper::AddField(Value* pJsonObj, Document* pDocument, const char* pCharFieldName, __int64 value)
{
	pJsonObj->AddMember<__int64>(pCharFieldName, value, pDocument->GetAllocator());
}

void JSONHelper::AddField(Value* pJsonObj, Document* pDocument, const char* pCharFieldName, unsigned __int64 value)
{
	pJsonObj->AddMember<unsigned __int64>(pCharFieldName, value, pDocument->GetAllocator());
}

void JSONHelper::AddField(Value* pJsonObj,Document* pDocument, const char* pCharFieldName, float value)
{
	pJsonObj->AddMember<float>(pCharFieldName, value, pDocument->GetAllocator());
}

void JSONHelper::AddField(Value* pJsonObj, Document* pDocument, const char* pCharFieldName, string value)
{
	Value field;
	field.SetString(value.c_str(), value.length(), pDocument->GetAllocator());

	pJsonObj->AddMember(pCharFieldName, field, pDocument->GetAllocator());
}

void JSONHelper::AddField(Value* pJsonObj, Document* pDocument, const char* pCharFieldName, Value value)
{
	pJsonObj->AddMember(pCharFieldName, value, pDocument->GetAllocator());
}

void JSONHelper::AddField(Value* pJsonObj, Document* pDocument, const char* pCharFieldName, vector<string> value)
{
	Value field(kArrayType);

	for (vector<string>::iterator it = value.begin(); it != value.end(); ++it)
	{
		Value element;
		element.SetString((*it).c_str(), (*it).length(), pDocument->GetAllocator());

		field.PushBack(element, pDocument->GetAllocator());
	}

	pJsonObj->AddMember(pCharFieldName, field, pDocument->GetAllocator());
}

void JSONHelper::AddField(Value* pJsonObj, Document* pDocument, const char* pCharFieldName, map<int, string> value)
{
	Value field(kObjectType);
	field.SetObject();
	for (map<int, string>::iterator itr = value.begin(); itr != value.end(); itr++)
	{
		//	key
		Value k;
		k.SetString(to_string(itr->first).c_str(), to_string(itr->first).length(), pDocument->GetAllocator());

		//	value
		Value v;
		v.SetString(itr->second.c_str(), itr->second.length(), pDocument->GetAllocator());

		field.AddMember(k, v, pDocument->GetAllocator());
	}
	pJsonObj->AddMember(pCharFieldName, field, pDocument->GetAllocator());
}

void JSONHelper::AddField(Value* pJsonObj, Document* pDocument, const char* pCharFieldName, IJSONObjectConvertible* pValue)
{
	pJsonObj->AddMember(pCharFieldName, pValue->GetJSONObject(pDocument), pDocument->GetAllocator());
}
#pragma endregion


#pragma region GetField
bool JSONHelper::GetField(Value* pJsonObj, string strFieldName, __int32* pValue)
{
	if (!pJsonObj->HasMember(strFieldName.c_str()))
	{
		//Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
		return false;
	}

	if (!(*pJsonObj)[strFieldName.c_str()].IsInt())
	{
		//Debug.LogWarning("Data type is invalid! It's not Int");
		return false;
	}

	*pValue = (*pJsonObj)[strFieldName.c_str()].GetInt();

	return true;
}

bool JSONHelper::GetField(Value* pJsonObj, string strFieldName, __int64* pValue)
{
	if (!pJsonObj->HasMember(strFieldName.c_str()))
	{
		//Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
		return false;
	}

	if (!(*pJsonObj)[strFieldName.c_str()].IsInt64())
	{
		//Debug.LogWarning("Data type is invalid! It's not Int64");
		return false;
	}

	*pValue = (*pJsonObj)[strFieldName.c_str()].GetInt64();

	return true;
}

bool JSONHelper::GetField(Value* pJsonObj, string strFieldName, unsigned __int64* pValue)
{
	if (!pJsonObj->HasMember(strFieldName.c_str()))
	{
		//Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
		return false;
	}

	if (!(*pJsonObj)[strFieldName.c_str()].IsUint64())
	{
		//Debug.LogWarning("Data type is invalid! It's not Uint64");
		return false;
	}

	*pValue = (*pJsonObj)[strFieldName.c_str()].GetUint64();

	return true;
}

bool JSONHelper::GetField(Value* pJsonObj, string strFieldName, float* pValue)
{
	if (!pJsonObj->HasMember(strFieldName.c_str()))
	{
		//Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
		return false;
	}

	if ((*pJsonObj)[strFieldName.c_str()].IsDouble())
	{
		*pValue = (*pJsonObj)[strFieldName.c_str()].GetDouble();
	}
	else if ((*pJsonObj)[strFieldName.c_str()].IsInt())
	{
		*pValue = (*pJsonObj)[strFieldName.c_str()].GetInt();
	}
	else
	{
		//Debug.LogWarning("Data type is invalid! It's not Float");
		return false;
	}

	return true;
}

bool JSONHelper::GetField(Value* pJsonObj, string strFieldName, string* pValue)
{
	if (!pJsonObj->HasMember(strFieldName.c_str()))
	{
		//Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
		return false;
	}

	if (!(*pJsonObj)[strFieldName.c_str()].IsString())
	{
		//Debug.LogWarning("Data type is invalid! It's not String");
		return false;
	}

	*pValue = (*pJsonObj)[strFieldName.c_str()].GetString();

	return true;
}

bool JSONHelper::GetField(Value* pJsonObj, string strFieldName, list<__int32>* pValue)
{
	if (!pJsonObj->HasMember(strFieldName.c_str()))
	{
		//Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
		return false;
	}

	if (!(*pJsonObj)[strFieldName.c_str()].IsArray())
	{
		//Debug.LogWarning("Data type is invalid! It's not Array");
		return false;
	}

	const Value& values = (*pJsonObj)[strFieldName.c_str()];
	for (SizeType i = 0; i < values.Size(); i++)
	{
		(*pValue).push_back(values[i].GetInt());
	}

	return true;
}

bool JSONHelper::GetField(Value* pJsonObj, string strFieldName, vector<string>* pValue)
{
	if (!pJsonObj->HasMember(strFieldName.c_str()))
	{
		//Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
		return false;
	}

	if (!(*pJsonObj)[strFieldName.c_str()].IsArray())
	{
		//Debug.LogWarning("Data type is invalid! It's not Array");
		return false;
	}

	const Value& values = (*pJsonObj)[strFieldName.c_str()];
	for (SizeType i = 0; i < values.Size(); i++)
	{
		(*pValue).push_back(values[i].GetString());
	}

	return true;
}

bool JSONHelper::GetField(Value* pJsonObj, string strFieldName, map<int, string>* pValue)
{
	if (!pJsonObj->HasMember(strFieldName.c_str()))
	{
		//Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
		return false;
	}

	if (!(*pJsonObj)[strFieldName.c_str()].IsObject())
	{
		//Debug.LogWarning("Data type is invalid! It's not Array");
		return false;
	}

	const Value& values = (*pJsonObj)[strFieldName.c_str()];
	for (Value::ConstMemberIterator itr = values.MemberBegin(); itr != values.MemberEnd(); ++itr)
	{
		(*pValue)[atoi(itr->name.GetString())] = itr->value.GetString();
	}

	return true;
}

bool JSONHelper::GetField(Value* pJsonObj, string strFieldName, IJSONObjectConvertible* pValue)
{
	if (!pJsonObj->HasMember(strFieldName.c_str()))
	{
		//Debug.LogWarning("JSONObject does not have field, field name : " + strFieldName);
		return false;
	}

	if (!(*pJsonObj)[strFieldName.c_str()].IsObject())
	{
		//Debug.LogWarning("Data type is invalid! It's not Array");
		return false;
	}

	pValue->SetJSONObject(&((*pJsonObj)[strFieldName.c_str()]));

	return true;
}
#pragma endregion